go/control/grpc.go

```diff
@@ -0,0 +1,53 @@
+package control
+
+import (
+	"context"
+
+	"github.com/oasislabs/oasis-core/go/client"
+	"github.com/oasislabs/oasis-core/go/common/grpc"
+	pb "github.com/oasislabs/oasis-core/go/grpc/control"
+	"github.com/oasislabs/oasis-core/go/worker/registration"
+)
+
+var (
+	_ pb.ControlServer = (*grpcServer)(nil)
+)
+
+type grpcServer struct {
+	registration *registration.Worker
+	nodeClient   *client.Client
+}
+
+func (s *grpcServer) RequestShutdown(ctx context.Context, req *pb.ShutdownRequest) (*pb.ShutdownResponse, error) {
+	s.registration.ScheduleShutdown()
```

Comment:
A quick thought (not sure if this was already considered):

I think it would make a bit more sense if `ScheduleShutdown` was a method of `cmd/node/node.go`. 
That method would stop the registration worker and wait that node is un-registered and then also stop itself.

I think that would also avoid the need for passing in the `ShutdownCallback` to the registration worker (which "feels" wrong) and make this more extendable (in case in future we'd like to do more things on shutdown request).

Reply:
Yeah it would be a bit strange to have all this logic in the node as it needs to interact with the registry which is something that the registration worker does, so I think it belongs there.

Instead of passing an arbitrary callback you could just pass a stoppable interface (one with a `Stop()` method) so you could pass the node instance directly and avoid importing the node command. That would be slightly cleaner IMO.

Reply:
> Yeah it would be a bit strange to have all this logic in the node as it needs to interact with the registry which is something that the registration worker does, so I think it belongs there.

That logic could stay in the registration worker, but would be under the registration worker `Stop()`  method. I don't think it is correct to tie the "stopping of registration worker" with shutting down the Node. For example we might want to support stopping the registration worker without shutting down the node in future (while the other way: shutting down a node gracefully, should always wait for the registration worker to stop and de-register), that's why i think the `ScheduleShutdown` should be a `Node` method, that calls `registrationWorker.Stop()` (which would contain the logic to also wait that node is deregistered) and then shuts itself down.

Reply:
> that's why i think the ScheduleShutdown should be a Node method, that calls registrationWorker.Stop() (which would contain the logic to also wait that node is deregistered) and then shuts itself down.

Ah, yeah that makes sense. So the change would be minor -- that instead of calling a callback it would only signal that it has stopped. And then an external caller (e.g. the Node) would wait on that event and shut itself down.

I like this better and it should be a simple change.